cpp/src/fil/treelite_import.cu

```diff
@@ -349,69 +356,53 @@ int tree2fil(std::vector<fil_node_t>& nodes,
-    int cur           = top.second;
-    stack.pop();
-
-    while (!tree.IsLeaf(node_id)) {
+  int sparse_index = 1;
+  walk_tree<int>(
+    tree,
+    [&](int node_id, int fil_node_id) {
       // reserve space for child nodes
       // left is the offset of the left child node relative to the tree root
       // in the array of all nodes of the FIL sparse forest
-      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * cur + 1 : sparse_index - root;
+      int left = node_traits<fil_node_t>::IS_DENSE ? 2 * fil_node_id + 1 : sparse_index;
       sparse_index += 2;
       conversion_state<fil_node_t> cs = tl2fil_inner_node<fil_node_t>(
         left, tree, node_id, cat_sets, &cat_sets->bit_pool_offsets[tree_idx]);
-      nodes[root + cur] = cs.node;
-      // push child nodes into the stack
-      stack.push(pair_t(cs.tl_right, left + 1));
-      // stack.push(pair_t(tl_left, left));
-      node_id = cs.tl_left;
-      cur     = left;
-    }
-
-    // leaf node
-    nodes[root + cur] = fil_node_t({}, {}, 0, false, true, false, 0);
-    tl2fil_leaf_payload(
-      &nodes[root + cur], root + cur, tree, node_id, forest_params, vector_leaf, leaf_counter);
-  }
-
+      nodes[root + fil_node_id] = cs.node;
+
+      return cs.swap_child_nodes ? std::tuple(left + 1, left) : std::tuple(left, left + 1);
+    },
+    [&](int node_id, int fil_node_id) {
+      nodes[root + fil_node_id] = fil_node_t({}, {}, 0, false, true, false, 0);
+      tl2fil_leaf_payload(&nodes[root + fil_node_id],
+                          root + fil_node_id,
+                          tree,
+                          node_id,
+                          forest_params,
+                          vector_leaf,
+                          leaf_counter);
+    });
   return root;
 }
 
 struct level_entry {
   int n_branch_nodes, n_leaves;
 };
-typedef std::pair<int, int> pair_t;
 // hist has branch and leaf count given depth
 template <typename T, typename L>
-inline void tree_depth_hist(const tl::Tree<T, L>& tree, std::vector<level_entry>& hist)
+inline void node_depth_hist(const tl::Tree<T, L>& tree, std::vector<level_entry>& hist)
 {
-  std::stack<pair_t> stack;  // {tl_id, depth}
-  stack.push({tree_root(tree), 0});
-  while (!stack.empty()) {
-    const pair_t& top = stack.top();
-    int node_id       = top.first;
-    int depth         = top.second;
-    stack.pop();
-
-    while (!tree.IsLeaf(node_id)) {
-      if (static_cast<std::size_t>(depth) >= hist.size()) hist.resize(depth + 1, {0, 0});
+  walk_tree<std::size_t>(
+    tree,
+    [&](int node_id, std::size_t depth) {
+      if (depth >= hist.size()) hist.resize(depth + 1, {0, 0});
       hist[depth].n_branch_nodes++;
-      stack.push({tree.LeftChild(node_id), depth + 1});
-      node_id = tree.RightChild(node_id);
-      depth++;
-    }
-
-    if (static_cast<std::size_t>(depth) >= hist.size()) hist.resize(depth + 1, {0, 0});
-    hist[depth].n_leaves++;
-  }
+      return std::tuple(depth + 1, depth + 1);
```

Comment:
```suggestion
      return {depth + 1, depth + 1};
```

Reply:
doesn't work. From what I've read, it shouldn't work in any return context, but I added this wherever the compiler let me sneak it in.
What is the type of such a thing anyway?